### ISSUE TITLE
Please add support for tuple patterns

### ISSUE THREAD

**Issue by scalablecory** [dotnet/roslyn#15899]
I would like to see `switch` support tuple syntax:
```
switch((x, y))
{
case (0, "foo"):
break;
case (1, "bar"):
break;
}
```

I would expect similar semantics to current `switch`: only constants can be used. Transformation would be similar to string-based switch: computing hash for each case at compile-time, `switch` on the value's hash, and compare value at the end to ensure a match.

**Comment by alrz** [comment]
Thus is pattern matching which is likely to be available in the next version.

https://github.com/dotnet/roslyn/blob/master/docs/Language%20Feature%20Status.md

**Comment by bondsbw** [comment]
The spec doesn't seem to make it clear whether this syntax is planned for `switch` statements.  See #15904.

**Comment by alrz** [comment]
It is one of the recursive patterns (along with property and positional patterns). You can find related issues in millstone 3.0.

**Comment by sab39** [comment]
This issue talks about using tuples like Constant patterns, but I noticed that tuple types are broken even as type patterns, as in
```c#
switch (o)
{
  case string myString: handleString(myString); break; // Works fine
  case (int, int) intPair: doSomethingElse(intPair); break; // Syntax error
  case ValueTuple<int,int> intPair: doSomethingElse(intPair); break; // This version works
}
```
Is this a bug or is it deliberately not supported?

**Comment by jcouv** [comment]
This is a language design issue, which we have been moving over to the csharplang repo.
There are different sorts of patterns which were not implemented in C# 7.0, but are tracked in the [patterns proposal](https://github.com/dotnet/csharplang/blob/master/proposals/patterns.md#positional-pattern), including recursive and positional patterns that you describe.

I'll close the present issue and let the discussion continue over on the new repo. Thanks
